File: src/exercise_action_cpp/src/move_to_point_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "base_interfaces/action/move_to_point.hpp"

#include <cmath>

constexpr double EPS = 1e-8;

using base_interfaces::action::MoveToPoint;
using std::placeholders::_1;
using std::placeholders::_2;

inline double get_distance(const double &x1, const double &y1,
                           const double &x2, const double &y2
) {
    return sqrt((x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 - y2));
}

inline bool is_equal(double a, double b) {
    return fabs(a - b) < EPS;
}

class MoveToPointServer : public rclcpp::Node {
    friend inline double get_distance(const double &x1, const double &y1, const double &x2, const double &y2);
    friend inline bool is_equal(double a, double b);

public:
    MoveToPointServer() : Node("move_to_point_server_cpp"), current_x_(0.0), current_y_(0.0) {
        RCLCPP_INFO(this->get_logger(), "动作服务端创建成功!");
        /*
            rclcpp_action::Server<ActionT>::SharedPtr create_server<ActionT, NodeT>(
                NodeT node, const std::string &name, rclcpp_action::Server<ActionT>::GoalCallback handle_goal, rclcpp_action::Server<ActionT>::CancelCallback handle_cancel, rclcpp_action::Server<ActionT>::AcceptedCallback handle_accepted
            )
        */
        server_ = rclcpp_action::create_server<MoveToPoint>(
            this,
            "move_to_point",
            std::bind(&MoveToPointServer::goal_callback, this, _1, _2),
            std::bind(&MoveToPointServer::cancel_callback, this, _1),
            std::bind(&MoveToPointServer::accepted_callback, this, _1)
        );
    }

private:
    /*
        std::function<GoalResponse(const GoalUUID &, std::shared_ptr<const typename ActionT::Goal>)>
    */
    rclcpp_action::GoalResponse goal_callback(
        const rclcpp_action::GoalUUID &uuid, std::shared_ptr<const MoveToPoint::Goal> goal
    ) {
        (void)uuid;

        double target_x = goal->target_x;
        double target_y = goal->target_y;
        double speed = goal->speed;

        if (!std::isfinite(target_x) ||
            !std::isfinite(target_y) ||
            !std::isfinite(speed)
        ) {
            RCLCPP_ERROR(this->get_logger(), "参数必须为有限数!");
            return rclcpp_action::GoalResponse::REJECT;
        }

        if (target_x < -10.0 || target_x > 10.0 ||
            target_y < -10.0 || target_y > 10.0
        ) {
            RCLCPP_ERROR(this->get_logger(), "目标坐标必须在[-10.0, 10.0]!");
            return rclcpp_action::GoalResponse::REJECT;
        }

        if (speed <= 0.0) {
            RCLCPP_ERROR(this->get_logger(), "速度必须为正数!");
            return rclcpp_action::GoalResponse::REJECT;
        } else if (speed > 1.5) {
            RCLCPP_ERROR(this->get_logger(), "速度太大!");
            return rclcpp_action::GoalResponse::REJECT;
        }

        if (abs(get_distance(target_x, target_y, current_x_, current_y_)) < 0.001) {
            RCLCPP_ERROR(this->get_logger(), "目标点距离当前位置太小!");
            return rclcpp_action::GoalResponse::REJECT;
        }

        RCLCPP_INFO(this->get_logger(), "参数全部合法!");
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    /*
        std::function<CancelResponse(std::shared_ptr<ServerGoalHandle<ActionT>>)>;
    */
    rclcpp_action::CancelResponse cancel_callback(
        std::shared_ptr<rclcpp_action::ServerGoalHandle<MoveToPoint>> goal_handle
    ) {
        (void)goal_handle;
        RCLCPP_INFO(this->get_logger(), "收到取消请求, 准备取消任务!");
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    /*
        std::function<void (std::shared_ptr<ServerGoalHandle<ActionT>>)>;
    */
    void accepted_callback(
        std::shared_ptr<rclcpp_action::ServerGoalHandle<MoveToPoint>> goal_handle
    ) {
        std::thread(std::bind(&MoveToPointServer::execute, this, goal_handle)).detach();
    }

    void execute(
        std::shared_ptr<rclcpp_action::ServerGoalHandle<MoveToPoint>> goal_handle
    ) {
        // 目标相关
        double target_x = goal_handle->get_goal()->target_x;
        double target_y = goal_handle->get_goal()->target_y;
        double speed = goal_handle->get_goal()->speed;

        // 结果相关
        auto result = std::make_shared<MoveToPoint::Result>();
        double traveled_distance = 0.0;

        // 反馈相关
        auto feedback = std::make_shared<MoveToPoint::Feedback>();
        double start_x = current_x_;
        double start_y = current_y_;
        double remaining_distance = get_distance(target_x, target_y, start_x, start_y);
        float progress = 0.0;

        // 计算相关
        double dx = target_x - start_x;
        double dy = target_y - start_y;
        double total_dist = get_distance(target_x, target_y, start_x, start_y);
        double direction_x = dx / total_dist;
        double direction_y = dy / total_dist;

        rclcpp::Rate rate(2.0);

        while (!is_equal(target_x, current_x_) || !is_equal(target_y, current_y_)) {
            double normal_step = speed * 0.5;
            double actual_step = std::min(normal_step, remaining_distance);

            rate.sleep();

            // 处理取消请求
            if (goal_handle->is_canceling()) {
                result->success = false;
                result->message = "任务已被取消!";
                result->final_x = current_x_;
                result->final_y = current_y_;
                result->traveled_distance = traveled_distance;

                RCLCPP_INFO(this->get_logger(), "====================");
                RCLCPP_INFO(this->get_logger(), "最终结果: ");
                RCLCPP_INFO(this->get_logger(), "success = %s", result->success ? "true" : "false");
                RCLCPP_INFO(this->get_logger(), "message = %s", result->message.c_str());
                RCLCPP_INFO(this->get_logger(), "final_x = %.2lf", result->final_x);
                RCLCPP_INFO(this->get_logger(), "final_y = %.2lf", result->final_y);
                RCLCPP_INFO(this->get_logger(), "traveled_distance = %.2lf", result->traveled_distance);
                RCLCPP_INFO(this->get_logger(), "====================");

                goal_handle->canceled(result);
                return;
            }

            // 相关参数的计算
            current_x_ += direction_x * actual_step;
            current_y_ += direction_y * actual_step;
            traveled_distance += actual_step;
            remaining_distance = std::max(total_dist - traveled_distance, 0.0);
            progress = std::min(traveled_distance / total_dist, 1.0);

            // 反馈参数的赋值
            feedback->current_x = current_x_;
            feedback->current_y = current_y_;
            feedback->remaining_distance = remaining_distance;
            feedback->progress = progress;

            RCLCPP_INFO(this->get_logger(), "====================");
            RCLCPP_INFO(this->get_logger(), "连续反馈中...");
            RCLCPP_INFO(this->get_logger(), "当前x坐标: %.2lf", feedback->current_x);
            RCLCPP_INFO(this->get_logger(), "当前y坐标: %.2lf", feedback->current_y);
            RCLCPP_INFO(this->get_logger(), "剩余距离: %.2lf", feedback->remaining_distance);
            RCLCPP_INFO(this->get_logger(), "当前进度: %.2f", feedback->progress);
            RCLCPP_INFO(this->get_logger(), "====================");

            goal_handle->publish_feedback(feedback);
        }

        result->success = true;
        result->message = "任务已完成!";
        result->final_x = target_x;
        result->final_y = target_y;
        result->traveled_distance = total_dist;

        RCLCPP_INFO(this->get_logger(), "====================");
        RCLCPP_INFO(this->get_logger(), "最终结果: ");
        RCLCPP_INFO(this->get_logger(), "success = %s", result->success ? "true" : "false");
        RCLCPP_INFO(this->get_logger(), "message = %s", result->message.c_str());
        RCLCPP_INFO(this->get_logger(), "final_x = %.2lf", result->final_x);
        RCLCPP_INFO(this->get_logger(), "final_y = %.2lf", result->final_y);
        RCLCPP_INFO(this->get_logger(), "traveled_distance = %.2lf", result->traveled_distance);
        RCLCPP_INFO(this->get_logger(), "====================");

        goal_handle->succeed(result);
    }

    double current_x_;
    double current_y_;

    rclcpp_action::Server<MoveToPoint>::SharedPtr server_;

};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);

    auto server = std::make_shared<MoveToPointServer>();
    rclcpp::spin(server);

    rclcpp::shutdown();

    return 0;
}

File: src/exercise_action_py/exercise_action_py/move_to_point_server.py
import time
import math

import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer, GoalResponse, CancelResponse
from rclpy.action.server import ServerGoalHandle
from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup

from base_interfaces.action import MoveToPoint

def get_dist(x1: float, y1: float, x2: float, y2: float) -> float:
    return math.sqrt((x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 - y2))

def is_equal(a: float, b: float, eqs: float = 1e-6) -> bool:
    return abs(a - b) < eqs

class MoveToPointServer(Node):
    def __init__(self):
        super().__init__("move_to_point_server_py")
        self.get_logger().info("动作服务端创建成功(Python)!")

        self.current_x_ = 0.0
        self.current_y_ = 0.0

        self.callback_group_ = ReentrantCallbackGroup()
        self.server_ = ActionServer(
            self, MoveToPoint, "move_to_point",
            execute_callback = self.execute_callback,
            callback_group = self.callback_group_,
            goal_callback = self.goal_callback,
            cancel_callback = self.cancel_callback,
            handle_accepted_callback = self.handle_accepted_callback
        )

    def goal_callback(self, goal_request: MoveToPoint.Goal):
        target_x = goal_request.target_x
        target_y = goal_request.target_y
        speed = goal_request.speed

        if not math.isfinite(target_x) or not math.isfinite(target_y) or not math.isfinite(speed):
            self.get_logger().error("目标参数必须为有限数!")
            return GoalResponse.REJECT
        
        if target_x < -10.0 or target_x > 10.0 or target_y < -10.0 or target_y > 10.0:
            self.get_logger().error("目标位置的坐标必须在[-10.0, 10.0]!")
            return GoalResponse.REJECT
        
        if speed <= 0.0:
            self.get_logger().error("目标速度必须为正数!")
            return GoalResponse.REJECT
        elif speed > 1.5:
            self.get_logger().error("目标速度设置过大!")
            return GoalResponse.REJECT
        
        if get_dist(self.current_x_, self.current_y_, target_x, target_y) < 0.001:
            self.get_logger().error("目标与当前位置距离太小!")
            return GoalResponse.REJECT
        
        self.get_logger().info("参数全部合法!")
        return GoalResponse.ACCEPT

    def cancel_callback(self, cancel_request):
        self.get_logger().info("已收到取消请求, 准备取消任务!")
        return CancelResponse.ACCEPT

    def handle_accepted_callback(self, goal_handle):
        goal_handle.execute()

    def execute_callback(self, goal_handle: ServerGoalHandle):
        # 目标相关
        target_x = goal_handle.request.target_x
        target_y = goal_handle.request.target_y
        speed = goal_handle.request.speed

        # 结果相关
        result = MoveToPoint.Result()
        final_x = 0.0
        final_y = 0.0
        traveled_distance = 0.0

        # 反馈相关
        feedback = MoveToPoint.Feedback()
        remaining_distance = get_dist(self.current_x_, self.current_y_, target_x, target_y)
        progress = 0.0

        # 计算相关
        dx = target_x - self.current_x_
        dy = target_y - self.current_y_
        direction_x = dx / remaining_distance
        direction_y = dy / remaining_distance
        total_dist = get_dist(self.current_x_, self.current_y_, target_x, target_y)

        while not is_equal(target_x, self.current_x_) or not is_equal(target_y, self.current_y_):
            time.sleep(0.5)

            if goal_handle.is_cancel_requested:
                result.success = False
                result.message = "任务已取消!"
                result.final_x = self.current_x_
                result.final_y = self.current_y_
                result.traveled_distance= traveled_distance

                self.get_logger().info("====================")
                self.get_logger().info("最终结果")
                self.get_logger().info(f"success = {'True' if result.success else 'False'}")
                self.get_logger().info(f"message = {result.message}")
                self.get_logger().info(f"final_x = {result.final_x:.2f}")
                self.get_logger().info(f"final_y = {result.final_y:.2f}")
                self.get_logger().info(f"traveled_distance = {result.traveled_distance:.2f}")
                self.get_logger().info("====================")

                goal_handle.canceled()
                return result                

            normal_step = speed * 0.5
            actual_step = min(normal_step, remaining_distance)

            self.current_x_ += actual_step * direction_x
            self.current_y_ += actual_step * direction_y
            traveled_distance += actual_step
            remaining_distance = max((total_dist - traveled_distance), 0.0)
            progress = min(traveled_distance / total_dist, 1.0)

            feedback.current_x = self.current_x_
            feedback.current_y = self.current_y_
            feedback.remaining_distance = remaining_distance
            feedback.progress = progress

            self.get_logger().info("====================")
            self.get_logger().info("持续反馈中...")
            self.get_logger().info(f"当前x坐标: {feedback.current_x:.2f}")
            self.get_logger().info(f"当前y坐标: {feedback.current_y:.2f}")
            self.get_logger().info(f"剩余距离: {feedback.remaining_distance:.2f}")
            self.get_logger().info(f"当前进度: {feedback.progress:.2f}")
            self.get_logger().info("====================")

            goal_handle.publish_feedback(feedback)

        result.success = True
        result.message = "任务已完成!"
        result.final_x = target_x
        result.final_y = target_y
        result.traveled_distance= traveled_distance

        self.current_x_ = target_x
        self.current_y_ = target_y

        self.get_logger().info("====================")
        self.get_logger().info("最终结果")
        self.get_logger().info(f"success = {'True' if result.success else 'False'}")
        self.get_logger().info(f"message = {result.message}")
        self.get_logger().info(f"final_x = {result.final_x:.2f}")
        self.get_logger().info(f"final_y = {result.final_y:.2f}")
        self.get_logger().info(f"traveled_distance = {result.traveled_distance:.2f}")
        self.get_logger().info("====================")

        goal_handle.succeed()
        return result

def main():
    rclpy.init()

    server = MoveToPointServer()
    executor = MultiThreadedExecutor()
    executor.add_node(server)

    try:
        executor.spin()
    except KeyboardInterrupt:
        pass
    finally:
        executor.shutdown()
        server.destroy_node()
        
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == "__main__":
    main()